### ISSUE TITLE
SeLFiE: Do we really need to invent fresh names in syntactic sugars?

### ISSUE THREAD

**Issue by yutakang** [data61/PSL#117]
Due to the strict evaluation of SeLFiE stemming from the evaluation strategy of Standard ML, maybe we can safely assume that the arguments are normalised.

**Comment by yutakang** [comment]
No. 
Syntactic sugars are handled in `Eval_Syntactic_Sugar.ML`.
Normalisation happens in `Eval_Surface.ML` only after we `convert` syntactic sugars to the core syntax tree.

**Comment by yutakang** [comment]
This makes it difficult to implement `Is_Nth_Arg_Of` as a syntactic sugar because we cannot add 1.
Currently all atomic assertions have to return a boolean value for all domains (`path`, `number`, and `print`).

**Comment by yutakang** [comment]
We probably have to a genuine new atomic assertion `Is_Nth_Arg` in `Eval_Unode` and define `Is_Nth_Arg_Of` as a syntactic sugar of `Is_Nth_Arg` and `Is_An_Arg_Of`.

**Comment by yutakang** [comment]
We probably have to a genuine new atomic assertion `Is_Nth_Arg` in `Eval_Unode` and define `Is_Nth_Arg_Of` as a syntactic sugar of `Is_Nth_Arg` and `Is_An_Arg_Of`.

**Comment by yutakang** [comment]
Ah... developing `Is_Nth_Arg` in `Eval_Unode`  is not a good idea: It involves too much analysis about the structure of the syntax tree. 

Instead of introducing `Nth_Arg`, I introduce `Is_N_Plu_One_th_Child`.

**Comment by yutakang** [comment]
To answer the first question, the answer is yes. Because SeLFiE interpreter rewrite syntactic sugars to the corresponding core syntaxes before starts evaluation.

**Comment by yutakang** [comment]
I think we should introduce variable names just before Syntax.
In other words, we should have `from_Deep_to_Variable` instead of `from_Bound_to_Variable`.

Currently, we add names to variables using `from_Bound_to_Variable`.
But we are not passing name environments for variables when computing function bodies for lambda abstractions, for example. This is causing problems.

One way to fix this is to pass an environment at each transition, which causes code-clutters.
So, it is better to use de-bruijn indices just before `Eval_Surface`.
After `Eval_Surface`, we only introduce syntax sugars.

**Comment by yutakang** [comment]
Wait a second... We need at least the type of variables for quantifiers (`Eval_Quantifier`).

**Comment by yutakang** [comment]
But we do not really need string names for `Eval_Quantifier`, do we?

**Comment by yutakang** [comment]
Actually, not having variable names in `Eval_Quantifier` is even better: we can simplify `vname_n_subexpr_n_qtyp_to_apps` in  in `Eval_Quantifier`.

**Comment by yutakang** [comment]
Should we add variable names after or before `Eval_Deep`? 

Things to consider:
- We pass prints and natural numbers between two worlds.
- But if we use strict evaluation, we do not have to pass environments between two worlds: things are normalised before we pass them from an outer layer to an inner layer.
- If we introduce variable names before `Eval_Deep`, we have to convert de-Bruijn indices for both outer and inner worlds at the same time(?) and we cannot use the type class abstraction for this conversion to add variable names.

So, probably it is better to add variable names after `Eval_Deep`.

**Comment by yutakang** [comment]
We do not have to pass around environments between `Outer` and `Inner`. So, we should introduce variable names after `Eval_Deep`.

**Comment by yutakang** [comment]
No.... the other way around.
We should introduce `Deep` first, then variable names because we can introduce variable names as syntactic sugar.

**Comment by yutakang** [comment]
No... [look at this line](https://github.com/data61/PSL/blob/4290a16018c7fe73fb93e4c19a7b7136e3e80d32/SeLFiE/src/Interpreter/Eval_Deep.ML#L191).

We cannot have `Dive_In` outside like this: if we compute `Applies` like this, we prematurely normalise sub-terms.

We have to handle all steps as de-sugaring... we cannot have multiple modules that execute beta-reduction in SeLFiE.

We have to introduce `Dive_In` in `from_Parameter_With_Bool_to_Bound`.

**Comment by yutakang** [comment]
Duplicate of #123
